src/Microsoft.TestPlatform.Common/Filtering/FilterHelpers.cs

```diff
@@ -0,0 +1,231 @@
+    {
+        private const char EscapePrefix = '\\';    
+        private static readonly char[] SpecialCharacters = { '\\', '(', ')', '&', '|', '=', '!', '~' };
+
+        /// <summary>
+        /// Escapes a set of special characters for filter (%, (, ), &, |, =, !, ~) by replacing them with their escape sequences. 
+        /// </summary>
+        /// <param name="str">The input string that contains the text to convert.</param>
+        /// <returns>A string of characters with special characters converted to their escaped form.</returns>
+        public static string Escape(string str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            if (str.IndexOfAny(SpecialCharacters) < 0)
+            {
+                return str;
+            }
+
+            var builder = new StringBuilder();
+            for (int i = 0; i < str.Length; ++i)
+            {
+                var currentChar = str[i];
+                if (SpecialCharacters.Contains(currentChar))
+                {
+                    builder.Append(EscapePrefix);
+                }
+                builder.Append(currentChar);
+            }
+
+            return builder.ToString();
+        }
+
+        /// <summary>
+        /// Converts any escaped characters in the input filter string.
+        /// </summary>
+        /// <param name="str">The input string that contains the text to convert.</param>
+        /// <returns>A filter string of characters with any escaped characters converted to their unescaped form.</returns>
+        public static string Unescape(string str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            if (str.IndexOf(EscapePrefix) < 0)
+            {
+                return str;
+            }
+
+            var builder = new StringBuilder();
+            for (int i = 0; i < str.Length; ++i)
+            {
+                var currentChar = str[i];
+                if (currentChar == EscapePrefix)
+                {
+                    if (++i == str.Length || !SpecialCharacters.Contains(currentChar = str[i]))
+                    {
+                        // "\" should be followed by a special character.
+                        throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Resources.TestCaseFilterEscapeException, str));
+                    }
+                }
+
+                // Strictly speaking, string to be unescaped shouldn't contain any of the special characters, 
+                // other than being part of escape sequence, but we will ignore that to avoid additional overhead.
+
+                builder.Append(currentChar);
+            }
+
+            return builder.ToString();
+        }
+
+        internal static IEnumerable<string> TokenizeFilterExpressionString(string str)
+        {
+            if (str == null)
```

Comment:
This is to ensure the exact same behavior as before (with regex). we might be able to do further adjustment in `Parse` methods but I think it can be done in a separate PR.